pkg/asset/upgrade/bundle.go

```diff
@@ -0,0 +1,489 @@
+		"args":   ocCmd.Args,
+		"stdout": ocCmd.String(),
+		"stderr": ocCmd.String(),
+		"code":   ocCmd.ProcessState.ExitCode(),
+	}).Debug("Executed 'oc' command")
+	if err != nil {
+		return
+	}
+
+	// Extract from the release metadata JSON the digest of the release metadata image, so that
+	// we can later download it using the digest instead of the tag that we calculated earlier:
+	var releaseImageDigest string
+	err = g.jqTool.QueryBytes(
+		`.digest`,
+		ocStdout.Bytes(), &releaseImageDigest,
+	)
+	if err != nil {
+		return
+	}
+	releaseImageRef = fmt.Sprintf(
+		"%s/%s@%s",
+		bundleReleaseImageDomain, bundleReleaseImagePath, releaseImageDigest,
+	)
+
+	// Extract from the release metadata JSON the references of the rest of the images that are
+	// part of the release:
+	type ImageInfo struct {
+		Tag string `json:"tag"`
+		Ref string `json:"ref"`
+	}
+	var imageInfos []ImageInfo
+	err = g.jqTool.QueryBytes(
+		`[.references.spec.tags[] | {
+			"tag": .name,
+			"ref": .from.name
+		}]`,
+		ocStdout.Bytes(), &imageInfos,
+	)
+	if err != nil {
+		return
+	}
+
+	// Build the result map:
+	imageMap = map[string]string{
+		releaseImageTag: releaseImageRef,
+	}
+	for _, imageInfo := range imageInfos {
+		imageMap[imageInfo.Tag] = imageInfo.Ref
+	}
+
+	return
+}
+
+// downloadImages copies a set of images into the given registry server. The images are specified
+// with a map where the keys are the full image references and the values are the tags.
+func (g *bundleGenerator) downloadImages(ctx context.Context,
+	registryServer *registry.RegistryServer, imageMap map[string]string) error {
+	// Create a temporary directory to store the certificates and the pull secret that we will
+	// need to pass to the skopeo command:
+	tmpDir, err := os.MkdirTemp("", "*.skopeo")
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err := os.RemoveAll(tmpDir)
+		if err != nil {
+			g.logger.WithError(err).WithFields(logrus.Fields{
+				"dir": tmpDir,
+			}).Error("Failed to remove skopeo temporary directory")
+		}
+	}()
+
+	// Save the TLS certificate of the registry to the temporary directory, so that we can later
+	// pass it to the '--dest-cert-dir' file of the skopeo command.
+	certBytes, _ := registryServer.Certificate()
+	certDir := filepath.Join(tmpDir, "certs")
+	err = os.Mkdir(certDir, 0700)
+	if err != nil {
+		return err
+	}
+	certFile := filepath.Join(certDir, "tls.crt")
+	err = os.WriteFile(certFile, certBytes, 0400)
+	if err != nil {
+		return err
+	}
+
+	// Save the pull secret to a file in the temporary directory, so that we can later pass it
+	// to the '--src-authfile` of the skopeo command:
+	authFile := filepath.Join(tmpDir, "auth.json")
+	err = os.WriteFile(authFile, []byte(g.appCfg.Config.PullSecret), 0400)
+	if err != nil {
+		return err
+	}
+
+	// Download the images:
+	imageTags := maps.Keys(imageMap)
+	slices.Sort(imageTags)
+	registryAddr := registryServer.ListenAddr()
+	for i, srcTag := range imageTags {
```

Comment:
just to make sure I'm following. iiuc, the work here is something like this: https://github.com/openshift-kni/telco-ran-tools/blob/main/cmd/download.go#L482